src/seraphis_mocks/scan_context_async_mock.cpp

```diff
@@ -0,0 +1,797 @@
+            r = cryptonote::parse_and_validate_tx_base_from_blob(res.blocks[block_idx].txs[tx_idx].blob, tx);
+            THROW_WALLET_EXCEPTION_IF(!r, tools::error::wallet_internal_error,
+                "failed to parse tx blob at index " + std::to_string(tx_idx));
+
+            THROW_WALLET_EXCEPTION_IF(tx_idx >= block.tx_hashes.size(), tools::error::wallet_internal_error,
+                "unexpected number of tx hashes");
+
+            prepare_unscanned_legacy_transaction(block.tx_hashes[tx_idx],
+                std::move(tx),
+                get_total_output_count_before_tx(res.output_indices[block_idx].indices[1+tx_idx].indices),
+                unscanned_tx);
+        }
+    }
+}
+//-------------------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------------------
+bool AsyncScanContextLegacy::check_launch_next_task(const std::unique_lock<std::mutex> &pending_queue_lock)
+{
+    MDEBUG("Attempting to launch chunk task at " << m_scan_index.load(std::memory_order_relaxed)
+        << " (chunk_size_increment=" << m_max_chunk_size_hint << ")");
+
+    THROW_WALLET_EXCEPTION_IF(!pending_queue_lock.owns_lock(), tools::error::wallet_internal_error,
+        "Pending queue must be locked to check next task launch");
+
+    if (!m_scanner_ready.load(std::memory_order_relaxed))
+    {
+        MDEBUG("Pending queue is not available for use, no tasks can be launched");
+        return false;
+    }
+
+    if (m_end_scan_index != 0 && m_scan_index.load(std::memory_order_relaxed) >= m_end_scan_index)
+    {
+        MDEBUG("Scan tasks are scheduled to scan to chain tip, not launching another task");
+        return false;
+    }
+
+    if (m_num_pending_chunks.load(std::memory_order_relaxed) >= m_config.pending_chunk_queue_size)
+    {
+        MDEBUG("Pending queue is already at max capacity");
+        return false;
+    }
+
+    // We use a separate counter for scanning chunks so we don't overload memory.
+    // Continuously fetching chunks while the scanner is backstopped can overload memory.
+    if (m_num_scanning_chunks.load(std::memory_order_relaxed) >= m_config.pending_chunk_queue_size)
+    {
+        MDEBUG("Scanning queue is already at max capacity");
+        return false;
+    }
+
+    return true;
+}
+//-------------------------------------------------------------------------------------------------------------------
+void AsyncScanContextLegacy::fill_gap_if_needed(bool chunk_is_terminal_chunk,
+    const std::uint64_t &requested_chunk_size,
+    const sp::scanning::ChunkContext &chunk_context)
+{
+    if (!chunk_is_terminal_chunk)
+    {
+        // If chunk was smaller than requested, will need to fill the gap
+        const std::size_t chunk_size = sp::scanning::chunk_size(chunk_context);
+        const std::uint64_t gap = requested_chunk_size - chunk_size;
```

Comment:
Underflow shouldn't be possible here because of this in `parse_rpc_get_blocks`:

```
const std::uint64_t num_blocks = std::min((std::uint64_t)res.blocks.size(), chunk_request.requested_chunk_size);
unscanned_chunk_out.resize(num_blocks);
```

In the latest I throw if `requested_chunk_size < chunk_size` because otherwise the rest of the scanner will break if it actually saturates